observatory-dags/observatory/dags/telescopes/jstor.py

```diff
@@ -0,0 +1,396 @@
+    existing_labels = service.users().labels().list(userId='me').execute()['labels']
+    label_id = [label['id'] for label in existing_labels if label['name'] == label_name]
+    if label_id:
+        label_id = label_id[0]
+    else:
+        # create label
+        label_body = {
+            'name': label_name,
+            'messageListVisibility': 'show',
+            'labelListVisibility': 'labelShow',
+            'type': 'user'
+        }
+        result = service.users().labels().create(userId='me', body=label_body).execute()
+        label_id = result['id']
+    return label_id
+
+
+def message_has_label(message: dict, label_id: str) -> bool:
+    """ Checks if a message has the given label
+
+    :param message: A Gmail message
+    :param label_id: The label id
+    :return: True if the message has the label
+    """
+    label_ids = message['labelIds']
+    for label in label_ids:
+        if label == label_id:
+            return True
+
+
+def list_available_releases(service: Resource, publisher: str, processed_label_id: str) -> Dict[Pendulum, List[dict]]:
+    """ List the available releases by going through the messages of a gmail account and looking for a specific pattern.
+
+    If a message has been processed previously it has a specific label, messages with this label will be skipped.
+    The message should include a download url. The head of this download url contains the filename, from which the
+    release date and publisher can be derived.
+
+    :param service: Gmail service
+    :param publisher: Name of the publisher
+    :param processed_label_id: Id of the 'processed_reports' label
+    :return: Dictionary with release dates as key and reports info as value, where reports info is a list of country
+    and/or institution reports.
+    """
+
+    available_releases = {}
+    # Call the Gmail API
+    results = service.users().messages().list(userId='me', q='subject:"JSTOR Publisher Report Available"').execute()
+    for message_info in results['messages']:
+        message_id = message_info['id']
+        message = service.users().messages().get(userId='me', id=message_id).execute()
+
+        # check if message has label 'processed'
+        if message_has_label(message, processed_label_id):
+            continue
+
+        # get download url
+        download_url = None
+        message_data = base64.urlsafe_b64decode(message['payload']['body']['data'])
+        for link in BeautifulSoup(message_data, 'html.parser', parse_only=SoupStrainer('a')):
+            if link.text == 'Download Completed Report':
+                download_url = link['href']
+                break
+        if download_url is None:
+            raise AirflowException(f"Can't find download link for report in e-mail, message snippet: {message.snippet}")
+
+        # get filename from head
+        headers = create_headers(download_url)
+        time.sleep(20)
+        response = requests.head(download_url, headers=headers, allow_redirects=True)
```

Comment:
I don't think it is necessary to create headers for the Gmail API, it should still work without them?

Reply:
Yeah I think you're right. I added these at one point and thought it fixed the bot detection issues, but it seems to work without them as well, so probably a coincidence.